File: companion-plugin/core/src/main/java/ovh/aurumgg/companion/core/http/TokenAuth.java
package ovh.aurumgg.companion.core.http;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;

/** Проверка статического токена из заголовка Authorization. */
public final class TokenAuth {

    private final byte[] expected;

    public TokenAuth(String token) {
        this.expected = token.getBytes(StandardCharsets.UTF_8);
    }

    /**
     * Принимает «Bearer <token>» и голый токен. Сравнение постоянного времени —
     * иначе по времени ответа секрет можно подобрать посимвольно.
     */
    public boolean isAuthorized(String authorizationHeader) {
        if (authorizationHeader == null) return false;
        String candidate = authorizationHeader.strip();
        if (candidate.regionMatches(true, 0, "Bearer ", 0, 7)) {
            candidate = candidate.substring(7).strip();
        }
        if (candidate.isEmpty()) return false;
        return MessageDigest.isEqual(candidate.getBytes(StandardCharsets.UTF_8), expected);
    }
}

File: companion-mod/src/Aurum.Companion.Core/Http/TokenAuth.cs
using System;
using System.Text;

namespace Aurum.Companion.Core.Http
{
    /// <summary>
    /// Сверка предъявленного токена с настроенным.
    /// </summary>
    public static class TokenAuth
    {
        /// <summary>Вынимает токен из заголовка Authorization. Схема Bearer необязательна.</summary>
        public static string? Extract(string? authorizationHeader)
        {
            if (string.IsNullOrWhiteSpace(authorizationHeader)) return null;
            string value = authorizationHeader!.Trim();
            const string bearer = "Bearer ";
            if (value.StartsWith(bearer, StringComparison.OrdinalIgnoreCase))
            {
                value = value.Substring(bearer.Length).Trim();
            }
            return value.Length == 0 ? null : value;
        }

        /// <summary>
        /// Сравнение за постоянное время.
        /// </summary>
        /// <remarks>
        /// Обычное сравнение строк выходит на первом несовпавшем символе, и по
        /// времени ответа токен подбирается посимвольно. Сама длина секретом не
        /// является, поэтому её различие можно вернуть сразу.
        /// </remarks>
        public static bool Equals(string? expected, string? provided)
        {
            if (expected == null || provided == null) return false;
            byte[] a = Encoding.UTF8.GetBytes(expected);
            byte[] b = Encoding.UTF8.GetBytes(provided);
            if (a.Length != b.Length) return false;

            int diff = 0;
            for (int i = 0; i < a.Length; i++) diff |= a[i] ^ b[i];
            return diff == 0;
        }
    }
}
